File: bnbagent/apex/policy.py
"""Thin wrapper around ``OptimisticPolicy`` (APEX v1 reference policy).

Surface:

- ``dispute(jobId)``      — client-only, within dispute window.
- ``vote_reject(jobId)``  — whitelisted voter, post-dispute.
- Read helpers for window state, quorum, voter status, etc.

Note: the contract's "silence approves" design means voters can ONLY reject.
There is no ``voteApprove`` on-chain; jobs without dispute auto-approve when
``submittedAt + disputeWindow`` elapses.
"""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

from web3 import Web3
from web3.contract import Contract

from ..core.contract_mixin import ContractClientMixin
from ..wallets.wallet_provider import WalletProvider
from .types import Verdict


def _load_abi() -> list:
    abi_path = Path(__file__).parent / "abis" / "OptimisticPolicy.json"
    return json.loads(abi_path.read_text())


class PolicyClient(ContractClientMixin):
    """Low-level client for ``OptimisticPolicy``."""

    def __init__(
        self,
        web3: Web3,
        contract_address: str,
        wallet_provider: WalletProvider | None = None,
        *,
        abi: list | None = None,
    ) -> None:
        self.w3 = web3
        self.address = Web3.to_checksum_address(contract_address)
        self.contract: Contract = self.w3.eth.contract(
            address=self.address, abi=abi or _load_abi()
        )
        self._wallet_provider = wallet_provider
        self._account = wallet_provider.address if wallet_provider is not None else None

    # ----------------------------------------------------------------- writes

    def dispute(self, job_id: int) -> dict[str, Any]:
        """Client raises a dispute. MUST be within dispute window."""
        fn = self.contract.functions.dispute(job_id)
        return self._send_tx(fn)

    def vote_reject(self, job_id: int) -> dict[str, Any]:
        """Whitelisted voter casts a reject vote (one per voter per job)."""
        fn = self.contract.functions.voteReject(job_id)
        return self._send_tx(fn)

    # ------------------------------------------------------------------ views

    def check(self, job_id: int, evidence: bytes = b"") -> tuple[Verdict, bytes]:
        """Simulate the verdict the Router would see right now."""
        verdict_int, reason = self._call_with_retry(
            self.contract.functions.check(job_id, evidence)
        )
        return Verdict(verdict_int), reason

    def submitted_at(self, job_id: int) -> int:
        return self._call_with_retry(self.contract.functions.submittedAt(job_id))

    def disputed(self, job_id: int) -> bool:
        return self._call_with_retry(self.contract.functions.disputed(job_id))

    def reject_votes(self, job_id: int) -> int:
        return self._call_with_retry(self.contract.functions.rejectVotes(job_id))

    def has_voted(self, job_id: int, voter: str) -> bool:
        return self._call_with_retry(
            self.contract.functions.hasVoted(job_id, Web3.to_checksum_address(voter))
        )

    def is_voter(self, voter: str) -> bool:
        return self._call_with_retry(
            self.contract.functions.isVoter(Web3.to_checksum_address(voter))
        )

    def get_deliverable_url(self, job_id: int, *, hint_block: int | None = None) -> str | None:
        """Return the ``deliverable_url`` for a submitted job.

        Reads the ``JobInitialised`` event emitted by ``onSubmitted`` and
        parses ``optParams`` (JSON bytes) to extract ``deliverable_url``.
        Returns ``None`` if the event is not found or the field is absent.

        Args:
            hint_block: if the caller knows roughly when the job was submitted
                (e.g. the block number of the ``Disputed`` event), passing it
                allows the query to use a tight window and avoid RPC block-range
                limits.  When omitted a 5 000-block lookback is used.
        """
        _LOOKBACK = 5_000

        try:
            current_block = self.w3.eth.block_number
        except Exception:
            current_block = None

        if hint_block is not None:
            from_block = max(0, hint_block - _LOOKBACK)
            to_block = hint_block + 10
        elif current_block is not None:
            from_block = max(0, current_block - _LOOKBACK)
            to_block = "latest"
        else:
            from_block = 0
            to_block = "latest"

        try:
            logs = self.contract.events.JobInitialised().get_logs(
                from_block=from_block,
                to_block=to_block,
                argument_filters={"jobId": job_id},
            )
        except Exception as exc:
            logger.warning("[PolicyClient] get_deliverable_url(%s) event query failed: %s", job_id, exc)
            return None

        if not logs:
            return None

        raw: bytes = logs[0]["args"].get("optParams", b"")
        if not raw:
            return None
        try:
            params = json.loads(raw.decode("utf-8"))
            return params.get("deliverable_url") or None
        except Exception as exc:
            logger.warning("[PolicyClient] get_deliverable_url(%s) parse failed: %s", job_id, exc)
            return None

    def dispute_window(self) -> int:
        return self._call_with_retry(self.contract.functions.disputeWindow())

    def vote_quorum(self) -> int:
        return self._call_with_retry(self.contract.functions.voteQuorum())

    def active_voter_count(self) -> int:
        return self._call_with_retry(self.contract.functions.activeVoterCount())

    def admin(self) -> str:
        return self._call_with_retry(self.contract.functions.admin())

    def commerce(self) -> str:
        return self._call_with_retry(self.contract.functions.commerce())

    def router(self) -> str:
        return self._call_with_retry(self.contract.functions.router())

    # --------------------------------------------------- admin writes (owner)

    def add_voter(self, voter: str) -> dict[str, Any]:
        fn = self.contract.functions.addVoter(Web3.to_checksum_address(voter))
        return self._send_tx(fn)

    def remove_voter(self, voter: str) -> dict[str, Any]:
        fn = self.contract.functions.removeVoter(Web3.to_checksum_address(voter))
        return self._send_tx(fn)

    def set_quorum(self, new_quorum: int) -> dict[str, Any]:
        fn = self.contract.functions.setQuorum(new_quorum)
        return self._send_tx(fn)

File: typescript/src/erc8183/policy.ts
/**
 * Thin client around `OptimisticPolicy` (ERC-8183 reference policy).
 *
 * Surface:
 *
 * - `dispute(jobId)`    — client-only, within dispute window.
 * - `voteReject(jobId)` — whitelisted voter, post-dispute.
 * - Read helpers for window state, quorum, voter status, etc.
 *
 * Note: the contract's "silence approves" design means voters can ONLY
 * reject. There is no `voteApprove` on-chain; jobs without dispute
 * auto-approve when `submittedAt + disputeWindow` elapses.
 *
 * Port of `python/bnbagent/erc8183/policy.py`.
 */

import { type Hex, type PublicClient, getAddress, hexToString } from "viem";
import { optimisticPolicyAbi } from "../abis/optimisticPolicy.js";
import { ContractBase, type DecodedEventLog } from "../core/contractBase.js";
import type { Paymaster } from "../core/paymaster.js";
import { describeError } from "../core/txSender.js";
import { RpcRangeLimitError } from "../errors.js";
import {
  ERC8183_DISPUTE,
  ERC8183_VOTE_REJECT,
  type TxResult,
} from "../wallets/intents.js";
import type { WalletProvider } from "../wallets/walletProvider.js";
import type { Verdict } from "./types.js";

/** Options accepted by the {@link PolicyClient} constructor (beyond the wallet). */
export interface PolicyClientOpts {
  paymaster?: Paymaster | null;
  receiptTimeout?: number | null;
}

/** Options accepted by {@link PolicyClient.getDeliverableUrl}. */
export interface GetDeliverableUrlOpts {
  /** Block the `JobSubmitted`/`onSubmitted` call landed in, if known — narrows
   * the log scan to a tight `+-10` block window instead of a 1000-block
   * fallback. */
  hintBlock?: bigint;
}

const TIGHT_WINDOW_BLOCKS = 10n;
const FALLBACK_WINDOW_BLOCKS = 1_000n;

/**
 * Low-level client for `OptimisticPolicy`.
 *
 * `dispute`/`voteReject` go through `ContractBase`'s intent execution seam
 * (`executeIntent`). The owner-only admin ops (`addVoter`/`removeVoter`/
 * `setQuorum`) are deliberately NOT migrated to that seam — they go through
 * `sendTx` directly, mirroring the Python reference.
 */
export class PolicyClient extends ContractBase {
  constructor(
    client: PublicClient,
    contractAddress: string,
    walletProvider?: WalletProvider | null,
    opts?: PolicyClientOpts,
  ) {
    super({
      client,
      address: getAddress(contractAddress),
      abi: optimisticPolicyAbi,
      walletProvider: walletProvider ?? null,
      paymaster: opts?.paymaster ?? null,
      receiptTimeout: opts?.receiptTimeout ?? null,
    });
  }

  // ----------------------------------------------------------------- writes

  /** Client raises a dispute. MUST be within dispute window. */
  async dispute(jobId: bigint): Promise<TxResult> {
    return this.executeIntent({
      name: ERC8183_DISPUTE,
      kwargs: { jobId },
      call: {
        address: this.address,
        abi: this.abi,
        functionName: "dispute",
        args: [jobId],
      },
      description: "dispute job",
    });
  }

  /** Whitelisted voter casts a reject vote (one per voter per job). */
  async voteReject(jobId: bigint): Promise<TxResult> {
    return this.executeIntent({
      name: ERC8183_VOTE_REJECT,
      kwargs: { jobId },
      call: {
        address: this.address,
        abi: this.abi,
        functionName: "voteReject",
        args: [jobId],
      },
      description: "vote reject",
    });
  }

  // --------------------------------------------------------- admin writes

  /**
   * Owner-only admin op. Deliberately NOT run through the intent seam
   * (`executeIntent`) — always self-pays via `sendTx`.
   */
  async addVoter(voter: string): Promise<TxResult> {
    return this.sendTx({
      functionName: "addVoter",
      args: [getAddress(voter)],
    });
  }

  /** Owner-only admin op. Deliberately NOT run through the intent seam. */
  async removeVoter(voter: string): Promise<TxResult> {
    return this.sendTx({
      functionName: "removeVoter",
      args: [getAddress(voter)],
    });
  }

  /** Owner-only admin op. Deliberately NOT run through the intent seam. */
  async setQuorum(newQuorum: number): Promise<TxResult> {
    return this.sendTx({
      functionName: "setQuorum",
      args: [newQuorum],
    });
  }

  // ------------------------------------------------------------------ views

  /** Simulate the verdict the Router would see right now. */
  async check(jobId: bigint, evidence: Hex = "0x"): Promise<[Verdict, Hex]> {
    const [verdict, reason] = await this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "check",
        args: [jobId, evidence],
      }),
    );
    return [verdict as Verdict, reason as Hex];
  }

  async submittedAt(jobId: bigint): Promise<bigint> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "submittedAt",
        args: [jobId],
      }),
    );
  }

  async disputed(jobId: bigint): Promise<boolean> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "disputed",
        args: [jobId],
      }),
    );
  }

  async rejectVotes(jobId: bigint): Promise<number> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "rejectVotes",
        args: [jobId],
      }),
    );
  }

  async hasVoted(jobId: bigint, voter: string): Promise<boolean> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "hasVoted",
        args: [jobId, getAddress(voter)],
      }),
    );
  }

  async isVoter(voter: string): Promise<boolean> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "isVoter",
        args: [getAddress(voter)],
      }),
    );
  }

  async disputeWindow(): Promise<bigint> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "disputeWindow",
      }),
    );
  }

  async voteQuorum(): Promise<number> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "voteQuorum",
      }),
    );
  }

  /**
   * `voteQuorum` value snapshotted at `dispute()` time. Returns `0` if the
   * job has never been disputed. After `dispute`, the snapshot is the
   * threshold `check` will use, even if an admin later calls `setQuorum` —
   * protects pending disputes from retroactive admin adjustments.
   */
  async disputeQuorumSnapshot(jobId: bigint): Promise<number> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "disputeQuorumSnapshot",
        args: [jobId],
      }),
    );
  }

  async activeVoterCount(): Promise<number> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "activeVoterCount",
      }),
    );
  }

  async admin(): Promise<`0x${string}`> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "admin",
      }),
    );
  }

  async commerce(): Promise<`0x${string}`> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "commerce",
      }),
    );
  }

  async router(): Promise<`0x${string}`> {
    return this.callWithRetry(() =>
      this.client.readContract({
        address: this.address,
        abi: optimisticPolicyAbi,
        functionName: "router",
      }),
    );
  }

  /**
   * Return the `deliverable_url` for a submitted job.
   *
   * Reads the `JobInitialised` event emitted by `onSubmitted` and parses
   * `optParams` (JSON bytes) to extract `deliverable_url`. Returns `null`
   * if the event is not found or the field is absent. Throws
   * {@link RpcRangeLimitError} when the node rejects the log query with a
   * rate/range limit (`-32005`) — that is retryable, not proof of absence.
   *
   * Prefer calling the ERC8183 facade's `getDeliverableUrl`, which
   * auto-resolves `hintBlock` via Commerce's `JobSubmitted` event. If
   * called directly without `hintBlock`, a 1000-block fallback window is
   * used.
   */
  async getDeliverableUrl(
    jobId: bigint,
    opts?: GetDeliverableUrlOpts,
  ): Promise<string | null> {
    let currentBlock: bigint | null = null;
    try {
      currentBlock = await this.client.getBlockNumber();
    } catch {
      currentBlock = null;
    }

    let fromBlock: bigint;
    let toBlock: bigint | "latest";
    if (opts?.hintBlock !== undefined) {
      const hint = opts.hintBlock;
      fromBlock = hint > TIGHT_WINDOW_BLOCKS ? hint - TIGHT_WINDOW_BLOCKS : 0n;
      toBlock = hint + TIGHT_WINDOW_BLOCKS;
    } else if (currentBlock !== null) {
      fromBlock =
        currentBlock > FALLBACK_WINDOW_BLOCKS
          ? currentBlock - FALLBACK_WINDOW_BLOCKS
          : 0n;
      toBlock = "latest";
    } else {
      fromBlock = 0n;
      toBlock = "latest";
    }

    let logs: DecodedEventLog[];
    try {
      logs = await this.readEvents({
        eventName: "JobInitialised",
        fromBlock,
        toBlock,
        args: { jobId },
      });
    } catch (error) {
      const message = describeError(error);
      if (
        message.includes("-32005") ||
        message.toLowerCase().includes("limit exceeded")
      ) {
        // Rate/range-limited RPC is NOT "event not found" — surface a typed
        // retryable error instead of a null the caller would misread as a
        // genuinely absent deliverable.
        throw new RpcRangeLimitError(
          `JobInitialised scan for job ${jobId} hit the RPC range/rate limit; retry later`,
        );
      }
      console.warn(
        `[PolicyClient] getDeliverableUrl(${jobId}) event query failed: ${message}`,
      );
      return null;
    }

    if (logs.length === 0) {
      return null;
    }

    const raw = logs[0]?.args.optParams as Hex | undefined;
    if (!raw || raw === "0x") {
      return null;
    }
    try {
      const params = JSON.parse(hexToString(raw)) as {
        deliverable_url?: string;
      };
      return params.deliverable_url || null;
    } catch (error) {
      console.warn(
        `[PolicyClient] getDeliverableUrl(${jobId}) parse failed: ${describeError(error)}`,
      );
      return null;
    }
  }
}
